### PR TITLE
Add entrypoint to get QueryStateInfo by QueryId

### PR DESCRIPTION
**Pull request by rongrong**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 1]
```diff
@@ -78,4 +80,17 @@ public QueryStateInfoResource(
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()).map(ResourceGroupId::getRoot).map(rootResourceGroupInfos::get)))
                 .collect(toImmutableList());
     }
+
+    @GET
+    @Path("queryId/{queryId}")
+    public QueryStateInfo getQueryStateInfoByQueryId(@PathParam("queryId") String queryId)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: getQueryStateInfo is sufficient.

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 2]
```diff
@@ -53,7 +55,7 @@ public QueryStateInfoResource(
     }
 
     @GET
-    public List<QueryStateInfo> getQueryDiagnosticsByUser(@QueryParam("user") String user)
+    public List<QueryStateInfo> getQueryStateInfoByUser(@QueryParam("user") String user)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: getQueryStateInfos

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 3]
```diff
@@ -78,4 +80,17 @@ public QueryStateInfoResource(
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()).map(ResourceGroupId::getRoot).map(rootResourceGroupInfos::get)))
                 .collect(toImmutableList());
     }
+
+    @GET
+    @Path("queryId/{queryId}")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Why is this a path param and the other is a query param? Let's make them both path params.

Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Ya, we talked about the other one last time and I suggested to use query param because when we initially discussed about this we were saying we may want to allow multiple usernames. I don't know whether we still want to do that. If not, the other one could be a path param as well. Though there is a bit of overhead to change the client side to hit different api but it's not too bad. Another option is to have both as query param.

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Ah right.. let's make them both query params then.

**Comment by electrum** [comment]
Have Martin review this API. He has strong feelings about REST.

**Comment by rongrong** [comment]
@martint What's your opinion on this? If we are not interested in supporting multiple usernames per request, we can change everything to path. Or we can keep everything as parameter.

**Comment by martint** [comment]
`/v1/queryState` to represent the collection of all queries
`/v1/queryState?user=yyy` to filter the collection based on some criteria
`/v1/queryState/{queryId}` for addressing a specific query

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 4]
```diff
@@ -53,29 +54,44 @@ public QueryStateInfoResource(
...
+                    .filter(queryInfo -> Pattern.matches(user, queryInfo.getSession().getUser()))
+                    .collect(toImmutableList());
         }
-        List<QueryInfo> userQueries = queryManager.getAllQueryInfo().stream()
-                .filter(queryInfo -> Pattern.matches(user, queryInfo.getSession().getUser()))
-                .filter(queryInfo -> !queryInfo.getState().isDone())
-                .collect(toImmutableList());
 
-        Map<ResourceGroupId, ResourceGroupInfo> rootResourceGroupInfos = userQueries.stream()
+        queryInfos = queryInfos.stream().filter(queryInfo -> !queryInfo.getState().isDone()).collect(toImmutableList());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Add line breaks after every stream operation

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 5]
```diff
@@ -53,29 +54,44 @@ public QueryStateInfoResource(
...
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()),
                         queryManager.getQueryResourceGroup(queryInfo.getQueryId()).map(ResourceGroupId::getRoot).map(rootResourceGroupInfos::get)))
                 .collect(toImmutableList());
     }
+
+    @GET
+    @Path("{queryId}")
+    public QueryStateInfo getQueryStateInfo(@PathParam("queryId") String queryId)
+    {
+        QueryInfo queryInfo = queryManager.getQueryInfo(new QueryId(queryId));
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Does throwing NoSuchElementException cause 404? Because that's the right behavior from a REST endpoint.

**Comment by raghavsethi** [comment]
Assign back to me once comments are addressed.

File: presto-main/src/test/java/com/facebook/presto/server/TestQueryStateInfoResource.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+        Request request2 = preparePost()
+                .setUri(HttpUriBuilder.uriBuilderFrom(server.getBaseUrl()).replacePath("/v1/statement").build())
+                .setBodyGenerator(createStaticBodyGenerator("show catalogs", UTF_8))
+                .setHeader(PRESTO_USER, "user2")
+                .setHeader(PRESTO_SOURCE, "source")
+                .setHeader(PRESTO_CATALOG, "catalog")
+                .setHeader(PRESTO_SCHEMA, "schema")
+                .setHeader(PRESTO_CLIENT_INFO, "{\"clientVersion\":\"testVersion\"}")
+                .addHeader(PRESTO_SESSION, QUERY_MAX_MEMORY + "=1GB")
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/server/TestQueryStateInfoResource.java]: Are all these headers strictly necessary? If not please remove.

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 7]
```diff
@@ -53,29 +57,54 @@ public QueryStateInfoResource(
     }
 
     @GET
-    public List<QueryStateInfo> getQueryDiagnosticsByUser(@QueryParam("user") String user)
+    public List<QueryStateInfo> getQueryStateInfos(@QueryParam("user") String user)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: I missed this before, but these endpoints should declare that they produce JSON `@Produces(MediaType.APPLICATION_JSON)`

File: presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java [hunk 8]
```diff
@@ -53,29 +57,54 @@ public QueryStateInfoResource(
...
+    @Path("{queryId}")
+    public Response getQueryStateInfo(@PathParam("queryId") String queryId)
+    {
+        try {
+            QueryInfo queryInfo = queryManager.getQueryInfo(new QueryId(queryId));
+            return Response
+                    .ok()
+                    .entity(createQueryStateInfo(
+                            queryInfo,
+                            queryManager.getQueryResourceGroup(queryInfo.getQueryId()),
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/server/QueryStateInfoResource.java]: Pull this out into a variable

File: presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupInfo.java [hunk 9]
```diff
@@ -142,7 +156,6 @@ public ResourceGroupInfo createSingleNodeInfo()
                 getNumEligibleSubGroups(),
                 getMemoryUsage(),
                 getNumAggregatedRunningQueries(),
-                getNumAggregatedQueuedQueries(),
-                emptyList());
+                getNumAggregatedQueuedQueries());
```
Comment (raghavsethi) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupInfo.java]: This is an unrelated change, and adds very little. There is only one invocation of this function.

Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/resourceGroups/ResourceGroupInfo.java]: I added this because the newly added test complains. I think because originally I have the constructor with only one parameter without a json property annotation.

**Comment by raghavsethi** [comment]
Did you merge this?

**Comment by rongrong** [comment]
messed up the merge. Here's the new PR #7878
